src/tests/Interop/Swift/SwiftErrorHandling/SwiftErrorHandling.cs

```diff
@@ -0,0 +1,69 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Swift;
+using System.Text;
+using Xunit;
+
+public class ErrorHandlingTests
+{
+    private const string SwiftLib = "libSwiftErrorHandling.dylib";
+
+    [DllImport(SwiftLib, EntryPoint = "$s18SwiftErrorHandling05setMyB7Message5bytes6lengthySPys5UInt8VG_SitF")]
+    public static extern void SetErrorMessage(byte[] strBytes, int length);
+
+    [UnmanagedCallConv(CallConvs = new Type[] { typeof(CallConvSwift) })]
+    [DllImport(SwiftLib, EntryPoint = "$s18SwiftErrorHandling018conditionallyThrowB004willE0SiSb_tKF")]
+    public unsafe static extern nint conditionallyThrowError(bool willThrow, SwiftError* error);
+
+    [DllImport(SwiftLib, EntryPoint = "$s18SwiftErrorHandling05getMyB7Message4fromSPys5UInt8VGSgs0B0_p_tF")]
+    public static extern IntPtr GetErrorMessage(IntPtr handle);
+
+    [Fact]
+    public unsafe static void TestSwiftErrorThrown()
+    {
+        const string expectedErrorMessage = "Catch me if you can!";
+        SetErrorMessageForSwift(expectedErrorMessage);
+
+        SwiftError error;
+
+        // This will throw an error
+        conditionallyThrowError(true, &error);
+        Assert.True(error.Value != IntPtr.Zero, "A Swift error was expected to be thrown.");
+
+        string errorMessage = GetErrorMessageFromSwift(error);
+        Assert.True(errorMessage == expectedErrorMessage, string.Format("The error message retrieved from Swift does not match the expected message. Expected: {0}, Actual: {1}", expectedErrorMessage, errorMessage));
+    }
+
+    [Fact]
+    public unsafe static void TestSwiftErrorNotThrown()
+    {
+        const string expectedErrorMessage = "Catch me if you can!";
+        SetErrorMessageForSwift(expectedErrorMessage);
+
+        SwiftError error;
+
+        // This will not throw an error
+        int result = (int)conditionallyThrowError(false, &error);
+
+        Assert.True(error.Value == IntPtr.Zero, "No Swift error was expected to be thrown.");
+        Assert.True(result == 42, "The result from Swift does not match the expected value.");
+    }
+    
+    private static void SetErrorMessageForSwift(string message)
+    {
+        var messageBytes = Encoding.UTF8.GetBytes(message + char.MinValue); // Add null terminator
+        SetErrorMessage(messageBytes, messageBytes.Length);
```

Comment:
It would be more efficient and simpler to let the P/Invoke marshalling take care of the conversion to zero-terminated string and then construct the string using the constructor that takes UTF8 zero-terminated string on the Swift side: https://developer.apple.com/documentation/swift/string/init(validatingutf8:)-208fn

Reply:
Is there a P/Invoke marshalling mechanism that could take care of the conversion to zero-terminated string when receiving data from Interop call? I ran into a problem that string data from Swift wasn't zero-terminated and `PtrToStringUni` expected zero-terminated string. I solved it by adding zero-terminator on the Swift side (https://github.com/dotnet/runtime/pull/94764/commits/97a74cc2843028a7679ec8e67193b8f61b32c7ba) but there might be a cleaner solution.

Reply:
There is no magic to help in this case. You would have to return the length of the string in an extra `out` length parameter. 

I think it is fine to return the zero terminated like you have done it.

Reply:
On the other hand, it would be nice for both get... and set... methods to be symmetric. set... passes the length in explicitly, so it would be nice for get... to return it explicitly as well.

Reply:
That sounds good. Do you know if `out` parameters work with interop or is there another way of returning multiple values? If not, it would probably be "cleaner" to set the message without passing the length and do some additional work on swift side to read the string data properly.

Reply:
Yes, `out int length` parameter will work with interop. I 

>  there another way of returning multiple values

Another way to pass in or return multiple values is via a struct that has the pointer and the length as fields. 

I would go with `out int length` parameter in this case.

Reply:
Thank you, the `out` parameter works as intended with interop (https://github.com/dotnet/runtime/pull/94764/commits/7103cfc19b5ef32120ddb74b9c03d68ad49da472).